### PR TITLE
TMB normalization changes

### PR DESCRIPTION
**Pull request by runjin326**
# TMB normalization changes

## Description

Currently, TMB in OMPARE is calculated by dividing the total number of mutations pass filter by the length of a default bed file: `xgen-exome-research-panel-targets_hg38_ucsc_liftover.100bp_padded.sort.merged.bed`. However, not all samples uses that particular panel. This update matches each sample with the bed files that they actually use and re-calculate the TMB score.

Fixes # (issue)
https://github.com/d3b-center/bixu-tracker/issues/928

## Type of review for this PR
- [ ] Scientific / output review
- [x] Code quality review

note: scientific/output review requires the lead scientists or analysts of the project to be more focus on the output of the PR where code quality review requires reviewer to make sure the PR is following the [contributing guideline](https://www.notion.so/d3b/contributing-721d172616b24ff395a64c18c58534f5) 

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

The codes were ran sequentially through the terminal to make sure they all run smoothly.

QC plots with the output of the scripts were generated to ensure the results are as expected. 
For TCGA samples, I calculated the expected TMB scores using the TMB scores from current adult TMB file that is used in OMPARE, the bed length of default bed file in OMPARE and the bed length of the matching bed file for each file (assuming mutations are mainly in the matching bed regions). I visualize both by disease type (`QC/tcga_expected_vs_current_by_disease.png`) and by bed file (`QC/tcga_expected_vs_current_by_bed.png`). Basically, the ratio is around 1 for most of them but for samples using tcga_6k_genes_targetIntervals.bed, the TMB scores seemed to be lower than expected.

For PBTA results, I noticed that the TMB for the pediatrics are significantly lower than what we have currently in OMPARE, demonstrated in the figure `QC/in_use_vs_corrected_pbta.png`. To investigate that, I calculated the expected TMB based on bed lengths and compared to the previous bed-corrected TMB calculation. This analysis gives higher TMB score than previous bed-corrected TMB calculation, which is expected since we used mutect2 (vs. consensus), shown in `QC/corrected_vs_previous_corrected_pbta.png`. And then I also compared TMB score from this analysis to the expected TMB score (based on bed length differences) and it showed that they are comparable (shown in `QC/corrected_vs_expected_TMBscore_pbta.png and QC/corrected_vs_expected_TMB_pbta_by_bed.png`). After the discussion with @aadamk, we decided for the PBTA samples, we will still use `xgen-exome-research-panel-targets_hg38_ucsc_liftover.100bp_padded.sort.merged.bed` for consistency. 

The final plots for PNOC008 patients were used as additional QC for the script.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings
- [ ] I have emojis in my PR and commits 😉

## Questions for reviewers:
I have included the codes as a separate folder `tmb_calculation`, which is not integrated in the current OMPARE structure (just so that I don't mess up any of the current runs).  I can definitely make changes to make it compatible with the current structure. And I think for integrating this part to current OMPARE, we only need to integrate the two results: 
`../results/TCGA_not_in_pbta_diseasetypes_and_samples_TMBscores.txt` 
`../results/TCGA_in_pbta_diseasetypes_and_samples_TMBscores.txt`
and the bed file (for PNOC008 patient TMB calculation): 
`../references/ashion_confidential_exome_v2_2nt_pad.Gh38.bed`. The PBTA TMB scores should not have changed. 
I have pushed those files for review as well.

### CODE REVIEW

**Comment by komalsrathi** [comment]
@runjin326 I am starting to review this now. Will review each file - add comments/suggestions and when I am done I will let you know so you can make all the changes (if needed) in one go.

File: tmb_normalization/code/01-download-tcga-mutect2.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+# folder and put in `maf_files`
+
+import requests
+import json
+import pandas as pd
+import re
+import os
+
+# read in the manifest as a dataframe and subset the id column
+maf_manifest = pd.read_csv('../results/tcga_not_in_pbta_maf_manifest.tsv', sep='\t')
```
Comment (komalsrathi) [inline tmb_normalization/code/01-download-tcga-mutect2.py:12]: would be nice to have a way to add the `.git` root directory and then follow the sub-directories. That way this script can be run from any folder within OMPARE

**Comment by komalsrathi** [comment]
@runjin326 thanks for the thorough work! I went through all your scripts, looks like they are working just fine. I just made some minor changes to scripts 00-05 (directly added the commits to your branch) for reproducibility but they were all non-scientific changes. 

From what I can tell, scripts 00-05 need not be run unless major changes in the PBTA versions (we are using v19 and OpenPBTA will likely stop updating after v20). I will merge your changes _as is_ but will need to update a few things w.r.t. `06-tmb-plot.R` on my end to add the output files (generated after running 00-05) in the `data/reference` folder and integrate it with rest of OMPARE code to account for each incoming patient. But in short, no changes are needed from your side.

File: tmb_normalization/code/01-download-tcga-mutect2.py (hunk 1 above)
Comment (komalsrathi) [reply tmb_normalization/code/01-download-tcga-mutect2.py:12]: Not important - so I merged this _as is_ and we can revisit in future if needed (unlikely)

**Comment by runjin326** [comment]
@komalsrathi, thanks so much for reviewing these and committing the merge! :)

**Pull request merged**